apps/src/netsim/commands.js

```diff
@@ -0,0 +1,129 @@
+/* jshint
+ funcscope: true,
+ newcap: true,
+ nonew: true,
+ shadow: false,
+ unused: true,
+
+ maxlen: 90,
+ maxparams: 3,
+ maxstatements: 200
+ */
+'use strict';
+
+require('../utils');
+
+var commands = module.exports;
+
+/**
+ * @enum {int}
+ */
+var commandState = {
+  NOT_STARTED: 0,
+  WORKING: 1,
+  SUCCESS: 2,
+  FAILURE: 3
+};
+
+/**
+ *
+ * @constructor
+ */
+var Command = commands.Command = function () {
+  this.status_ = commandState.NOT_STARTED;
+};
+
+Command.prototype.isStarted = function () {
+  return this.status_ !== commandState.NOT_STARTED;
+};
+
+Command.prototype.isFinished = function () {
+  return this.succeeded() || this.failed();
+};
+
+Command.prototype.succeeded = function () {
+  return this.status_ === commandState.SUCCESS;
+};
+
+Command.prototype.failed = function () {
+  return this.status_ === commandState.FAILURE;
+};
+
+Command.prototype.begin = function () {
+  this.status_ = commandState.WORKING;
+  this.onBegin_();
+};
+
+Command.prototype.succeed = function () {
+  if (!this.isStarted()) {
+    throw new Error("Command cannot succeed before it begins.");
+  }
+  this.status_ = commandState.SUCCESS;
+  this.onEnd_();
+};
+
+Command.prototype.fail = function () {
+  if (!this.isStarted()) {
+    throw new Error("Command cannot fail before it begins.");
+  }
+  this.status_ = commandState.FAILURE;
+  this.onEnd_();
+};
+
+Command.prototype.onBegin_ = function () {
+
+};
+
+
+Command.prototype.tick = function () {
+
+};
+
+Command.prototype.onEnd_ = function () {
+
+};
+
+/**
+ *
+ * @constructor
+ */
+var CommandSequence = commands.CommandSequence = function (commandList) {
+  Command.call(this);
+  this.commandList_ = commandList;
+};
+CommandSequence.inherits(Command);
+
+CommandSequence.prototype.onBegin_ = function () {
+  this.currentIndex_ = 0;
```

Comment:
Could declare and type annotate in CommandSequence constructor. Potentially s/`currentIndex_`/`currentCommandIndex_`